### PR TITLE
IGNITE-18545 Fixed task execution options propagation.

### PR DESCRIPTION
**Pull request by petrov-mg**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        /** */
+        public TaskNameChecker(String expName) {
+            this.expName = expName == null ? getClass().getName() : expName;
+        }
+
+        /** */
+        protected void checkName() {
+            assertEquals(expName, ses.getTaskName());
+        }
+
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java:258]: Unnecessary line break

File: modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        runAsync(() -> compute.call(new TestCallable(TestCallable.class.getName()))).get();
+
+        compute.call(new TestCallable(TEST_TASK_NAME));
+    }
+
+    /** */
+    @Test
+    public void testTaskExecutionOptionsReset() throws Exception {
+        check(TASK, (c, t) -> c.execute((ComputeTask)t, null));
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java:92]: Let's cast to `ComputeTask<Void, Void>` as in the code below.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+        }
+    }
+
+    /** */
+    private static <T> Collection<T> toList(T t) {
+        return t == null ? null : singletonList(t);
+    }
+
+    /** */
+    private interface ConsumerX<T> {
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java:285]: ComputationConsumer ?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java (hunk 1 above)
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java (hunk 3 above)
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java [hunk 4]
```diff
@@ -626,21 +540,10 @@ private <T, R> ComputeTaskInternalFuture<R> startTask(
...
-            thCtx.set(null);
-
-        if (map.get(TC_SKIP_AUTH) == null)
+        if (!opts.isAuthenticationDisabled())
             ctx.security().authorize(taskClsName, SecurityPermission.TASK_EXECUTE);
 
-        Long timeout = (Long)map.get(TC_TIMEOUT);
-
-        long timeout0 = timeout == null || timeout == 0 ? Long.MAX_VALUE : timeout;
+        long timeout0 = opts.timeout().orElse(Long.MAX_VALUE);
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java]: Fixed. I think it's better not to return optional in TaskExecutionOptions#timeout() at all.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java (hunk 2 above)
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/compute/TaskOptionsPropagationTest.java]: Just refactored tests a bit so dedicated enum with computation classes is no longer needed. Please take a look again

**Review by NSAmelchev — APPROVED** [review]

**Pull request merged**